### PR TITLE
ClusterAutoscaler: Add support for startupTaints and statusTaints

### LINKED ISSUE #10844: clusterAutoscaler: Add support for `statusTaints` option mapping to `--status-taints` flag
**How to categorize this issue?**
/area auto-scaling
/kind enhancement

**What would you like to be added**:
In a shoot's `spec.kubernetes.clusterAutoscaler` map a `statusTaints` option should be added, functioning very similarly to the `ignoreTaints` option.

**Why is this needed**:
The `--ignore-taints` flag mentioned in https://github.com/gardener/gardener/issues/5166 [is now deprecated](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#ignore-taints) in favour of `--startup-taints` and its semantics have changed accordingly, such that any nodes with such a taint are considered unready by the cluster autoscaler (CA). This can cause the CA to stop working entirely if there are too many nodes with such taints.
Access to the new flag `--status-taints` is needed in order to restore the original functionality where the CA ignores the taint, but it does not affect the CA's view of node readiness.
At the same time, it is probably a good idea to deprecate `ignoreTaints` and support `startupTaints`.

### PR DESCRIPTION
**Pull request by dhague**
Deprecate ignoreTaints

**How to categorize this PR?**
/area auto-scaling
/kind enhancement

**What this PR does / why we need it**:
The --ignore-taints flag mentioned in https://github.com/gardener/gardener/issues/5166 [is now deprecated](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#ignore-taints) in favour of --startup-taints and its semantics have changed accordingly, such that any nodes with such a taint are considered unready by the cluster autoscaler (CA). This can cause the CA to stop working entirely if there are too many nodes with such taints.
Access to the new flag --status-taints is needed in order to restore the original functionality where the CA ignores the taint, but it does not affect the CA's view of node readiness.

**Which issue(s) this PR fixes**:
Fixes #10844

**Special notes for your reviewer**:

**Release note**:
```feature user
shoot spec.kubernetes.clusterAutoscaler: Add support for startupTaints and statusTaints
```

### CODE REVIEW

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 1]
```diff
@@ -633,21 +633,31 @@ type ClusterAutoscaler struct {
...
+	// +optional
+	StartupTaints []string `json:"startupTaints,omitempty" protobuf:"bytes,10,opt,name=startupTaints"`
+	// StatusTaints specifies a list of taint keys to ignore in node templates when considering to scale a node group.
+	// Cluster Autoscaler internally treats nodes tainted with status taints as ready, but filtered out during scale up logic.
+	// +optional
+	StatusTaints []string `json:"statusTaints,omitempty" protobuf:"bytes,11,opt,name=statusTaints"`
+	// Deprecated: Ignore taints are now deprecated and treated as startup taints.
 	// IgnoreTaints specifies a list of taint keys to ignore in node templates when considering to scale a node group.
 	// +optional
-	IgnoreTaints []string `json:"ignoreTaints,omitempty" protobuf:"bytes,10,opt,name=ignoreTaints"`
```
Comment (oliver-goetz) [inline pkg/apis/core/v1beta1/types_shoot.go:638]: You must not change protobuf field numbers ([see docs](https://github.com/gardener/gardener/blob/master/docs/development/changing-the-api.md#changing-the-api-1)). Please keep it consistent and append the new fields.

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 2]
```diff
@@ -633,21 +633,31 @@ type ClusterAutoscaler struct {
...
 	MaxGracefulTerminationSeconds *int32 `json:"maxGracefulTerminationSeconds,omitempty" protobuf:"varint,9,opt,name=maxGracefulTerminationSeconds"`
+	// StartupTaints specifies a list of taint keys to ignore in node templates when considering to scale a node group.
+	// Cluster Autoscaler treats nodes tainted with startup taints as unready, but taken into account during scale up logic, assuming they will become ready shortly.
+	// +optional
+	StartupTaints []string `json:"startupTaints,omitempty" protobuf:"bytes,10,opt,name=startupTaints"`
+	// StatusTaints specifies a list of taint keys to ignore in node templates when considering to scale a node group.
+	// Cluster Autoscaler internally treats nodes tainted with status taints as ready, but filtered out during scale up logic.
+	// +optional
+	StatusTaints []string `json:"statusTaints,omitempty" protobuf:"bytes,11,opt,name=statusTaints"`
+	// Deprecated: Ignore taints are now deprecated and treated as startup taints.
```
Comment (oliver-goetz) [inline pkg/apis/core/v1beta1/types_shoot.go]: I don't think we should already deprecate this field until we drop support for Kubernetes 1.28. `IgnoreTaints` seems to be the only config option for Kubernetes <= v1.28.

File: pkg/apis/core/validation/shoot.go [hunk 3]
```diff
@@ -1140,8 +1140,16 @@ func ValidateClusterAutoscaler(autoScaler core.ClusterAutoscaler, fldPath *field
 		}
 	}
 
+	if startupTaints := autoScaler.StartupTaints; startupTaints != nil {
+		allErrs = append(allErrs, validateClusterAutoscalerStartupStatusIgnoreTaints(startupTaints, fldPath.Child("startupTaints"))...)
+	}
+
+	if statusTaints := autoScaler.StatusTaints; statusTaints != nil {
+		allErrs = append(allErrs, validateClusterAutoscalerStartupStatusIgnoreTaints(statusTaints, fldPath.Child("statusTaints"))...)
+	}
```
Comment (oliver-goetz) [inline pkg/apis/core/validation/shoot.go]: I briefly scanned the cluster-autoscaler and it seems that these new flags are available for Kubernetes >= v1.29 only (see the flags for [cluster-autoscaler v1.28](https://github.com/gardener/autoscaler/blob/8d9feda527ed620d5fb3195246db08472b273b05/cluster-autoscaler/main.go#L195)). Gardener still supports older versions, so the Kubernetes version should be verified too, when these new taints are used.

**Review by oliver-goetz — CHANGES_REQUESTED** [review]
Thanks for your PR 😄

File: pkg/apis/core/validation/shoot.go [hunk 4]
```diff
@@ -1966,11 +1974,11 @@ func validateKubeletConfigReserved(reserved *core.KubeletConfigReserved, fldPath
 
 var reservedTaintKeys = sets.New(v1beta1constants.TaintNodeCriticalComponentsNotReady)
 
-func validateClusterAutoscalerIgnoreTaints(ignoredTaints []string, fldPath *field.Path) field.ErrorList {
+func validateClusterAutoscalerStartupStatusIgnoreTaints(taints []string, fldPath *field.Path) field.ErrorList {
```
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
func validateClusterAutoscalerTaints(taints []string, fldPath *field.Path) field.ErrorList {
```

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go [hunk 5]
```diff
@@ -84,7 +84,7 @@ var _ = Describe("ClusterAutoscaler", func() {
 		configVerbosity                     int32 = 4
 		configMaxEmptyBulkDelete                  = ptr.To[int32](20)
 		configNewPodScaleUpDelay                  = &metav1.Duration{Duration: time.Second}
-		configIgnoreTaints                        = []string{"taint-1", "taint-2"}
+		configStartupStatusIgnoreTaints           = []string{"taint-1", "taint-2"}
```
Comment (rfranzke) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go]:
```suggestion
		configTaints           = []string{"taint-1", "taint-2"}
```

File: pkg/apis/core/types_shoot.go [hunk 6]
```diff
@@ -504,7 +504,14 @@ type ClusterAutoscaler struct {
...
 	// MaxGracefulTerminationSeconds is the number of seconds CA waits for pod termination when trying to scale down a node (default: 600).
 	MaxGracefulTerminationSeconds *int32
+	// StartupTaints specifies a list of taint keys to ignore in node templates when considering to scale a node group.
+	// Cluster Autoscaler treats nodes tainted with startup taints as unready, but taken into account during scale up logic, assuming they will become ready shortly.
+	StartupTaints []string
+	// StatusTaints specifies a list of taint keys to ignore in node templates when considering to scale a node group.
+	// Cluster Autoscaler internally treats nodes tainted with status taints as ready, but filtered out during scale up logic.
+	StatusTaints []string
 	// IgnoreTaints specifies a list of taint keys to ignore in node templates when considering to scale a node group.
+	// Ignore taints are now deprecated and treated as startup taints.
```
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]: We should forbid setting this field for Kubernetes 1.32 and higher `Shoot`s to make sure we can eventually get rid of it in the `Shoot` API. Please adapt the doc string and implement this kind of validation.

File: pkg/apis/core/validation/shoot.go [hunk 7]
```diff
@@ -1966,11 +1974,25 @@ func validateKubeletConfigReserved(reserved *core.KubeletConfigReserved, fldPath
...
+	}
+
 	allErrs := field.ErrorList{}
+
+	supported, err := optionVersionRanges[option].Contains(version)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child(option), option, err.Error()))
+	} else if !supported {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child(option), "not supported in Kubernetes version "+version))
+	}
```
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go:1991]: Neat approach :) I like it.

**Review by rfranzke — COMMENTED** [review]
/lgtm

**Pull request merged**